habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -167,75 +169,99 @@ def __init__(self, *args, config, sim: RearrangeSim, **kwargs):
         super().__init__(*args, config=config, sim=sim, **kwargs)
         self._delta_pos_limit = self._config.delta_pos_limit
         self._arm_joint_mask = self._config.arm_joint_mask
+        self._arm_joint_limit = self._config.arm_joint_limit
+        if self._arm_joint_limit is not None:
+            assert (
+                len(self._arm_joint_limit)
+                == self._config.arm_joint_dimensionality
+            )
 
     @property
     def action_space(self):
+        assert self._config.arm_joint_dimensionality == len(
+            self._config.arm_joint_mask
+        ) or self._config.arm_joint_dimensionality == int(
+            np.sum(self._config.arm_joint_mask)
+        )
         return spaces.Box(
             shape=(self._config.arm_joint_dimensionality,),
             low=-1,
             high=1,
             dtype=np.float32,
         )
 
-    def step(self, delta_pos, should_step=True, *args, **kwargs):
-        # clip from -1 to 1
-        delta_pos = np.clip(delta_pos, -1, 1)
-        delta_pos *= self._delta_pos_limit
+    def _get_processed_action(self, delta_pos, simulation_mode="dynamic"):
+        """Assign the delta pos actions into a correct joint location"""
+        processed_delta_pos = np.zeros(len(self._arm_joint_mask))
+        min_limit, max_limit = self.cur_articulated_agent.arm_joint_limits
 
-        mask_delta_pos = np.zeros(len(self._arm_joint_mask))
         src_idx = 0
         tgt_idx = 0
         for mask in self._arm_joint_mask:
             if mask == 0:
                 tgt_idx += 1
-                src_idx += 1
+                # Check if the effective size of action is the same as arm_joint_dimensionality
+                if self._config.arm_joint_dimensionality == len(
+                    self._config.arm_joint_mask
+                ):
+                    src_idx += 1
                 continue
-            mask_delta_pos[tgt_idx] = delta_pos[src_idx]
+            processed_delta_pos[tgt_idx] = delta_pos[src_idx]
+
+            # Set the new limits if needed
+            if self._arm_joint_limit is not None:
+                min_limit[tgt_idx] = self._arm_joint_limit[src_idx][0]
+                max_limit[tgt_idx] = self._arm_joint_limit[src_idx][1]
+
             tgt_idx += 1
             src_idx += 1
 
-        # Although habitat_sim will prevent the motor from exceeding limits,
+        # Clip the action. Although habitat_sim will prevent the motor from exceeding limits,
         # clip the motor joints first here to prevent the arm from being unstable.
-        min_limit, max_limit = self.cur_articulated_agent.arm_joint_limits
-        target_arm_pos = (
-            mask_delta_pos + self.cur_articulated_agent.arm_motor_pos
-        )
+        if simulation_mode == "dynamic":
+            cur_arm_pos = self.cur_articulated_agent.arm_motor_pos
+        elif simulation_mode == "kinematic":
```

Comment:
Does this action even work in kinematic mode? You are setting the motor target, not the joint position kinematically?